src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -116,9 +115,15 @@ public GeneratorDriver TryApplyEdits(Compilation compilation, out Compilation ou
             return BuildFinalCompilation(compilation, out outputCompilation, state, cancellationToken);
         }
 
-        public GeneratorDriver WithGeneratorProviders(ImmutableArray<GeneratorProvider> providers)
+        public GeneratorDriver WithGenerators(ImmutableArray<ISourceGenerator> generators)
         {
-            var newState = _state.With(providers: _state.Providers.AddRange(providers));
+            var newState = _state.With(generators: _state.Generators.AddRange(generators));
```

Comment:
>_state.Generators.AddRange(generators) [](start = 51, length = 38)

It looks like the common pattern for `With` methods that take collections is to replace the existing collection rather than add to it.

Also, are `WithGenerators()` and `RemoveGenerator()` needed? What are the scenarios where these methods are used?

Reply:
It does replace them, it just replaces them with the existing collection + the extra one added. 

The other 'with' methods are used by the IDE. For instance when a user adds a new generator to the compilation; you can get the driver with the new generator without needing to recreate all the previous state. A lot of the complexity comes from needing to work in the IDE :/

---
In reply to: [387250370](https://github.com/dotnet/roslyn/pull/42030#discussion_r387250370) [](ancestors = 387250370)